app/src/lib/status-parser.ts

```diff
@@ -47,10 +49,12 @@ export function parsePorcelainStatus(
   // containing special characters are not specially formatted; no quoting or
   // backslash-escaping is performed.
 
-  const fields = output.split('\0')
+  const tokens = output.split('\0')
+  const queue = new Deque(tokens)
```

Comment:
@j-f1 I avoided this sort of approach (no need to push onto a new array first, the `split` gives us an array we own):

```ts
const fields = output.split('\0')
fields.reverse()

...

while ((field = queue.pop())) {
```

for a couple of reasons:

 - the algorithm is intended to work from beginning to end, and inverting these semantics might be confusing for future people reading the source
 - I wasn't sure how poorly `Array.reverse` would perform when dealing with an array of 100000 elements or so - this might perform better than calling `Array.shift` to consume the array, but I suspect we'd be paying the O(N) cost up front for large status outputs

A quick test of this in the Chrome dev tools:

```js
let arr = []

for (i = 0; i < 1000000; i++) {
  arr.push(i.toString())
}

console.time('reverse')

arr.reverse()

console.timeEnd('reverse')
```

Gives me a timing of 20-25ms - much improved, but given the worst case time of `parsePorcelainStatus` was below 80ms (and did a lot more) I suspect this approach still comes out on top. Happy to be proven wrong though.